src/System.CommandLine/Parsing/ParseResultVisitor.cs

```diff
@@ -34,19 +34,27 @@ internal class ParseResultVisitor : SyntaxVisitor
             _tokenizeResult = tokenizeResult;
             _rawInput = rawInput;
 
-            _unparsedTokens = new List<string>();
-            if (unparsedTokens?.Count > 0)
+            var unparsedTokensCount = unparsedTokens?.Count ?? 0;
+            _unparsedTokens = unparsedTokensCount == 0? new List<string>(): new List<string>(unparsedTokensCount);
+            if (unparsedTokensCount > 0)
             {
-                _unparsedTokens.AddRange(unparsedTokens.Select(t => t.Value));
+                foreach (var unparsedToken in unparsedTokens)
+                {
+                    _unparsedTokens.Add(unparsedToken.Value);
+                }
             }
 
-            _unmatchedTokens = new List<string>();
-            if (unmatchedTokens?.Count > 0)
+            var unmatchedTokensCount = unmatchedTokens?.Count ?? 0;
+            _unmatchedTokens = unmatchedTokensCount == 0 ? new List<string>() : new List<string>(unmatchedTokensCount);
+            if (unmatchedTokensCount > 0)
             {
-                _unmatchedTokens.AddRange(unmatchedTokens.Select(t => t.Value));
+                foreach (var unmatchedToken in unmatchedTokens)
+                {
+                    _unmatchedTokens.Add(unmatchedToken.Value);
+                }
             }
 
-            _errors = new List<ParseError>();
+            _errors = new List<ParseError>(2);
```

Comment:
Shouldn't this be?
```suggestion
            _errors = new List<ParseError>(_tokenizeResult.Errors.Count + parseErrors.Count);
```

Reply:
Good call! Missed that and the allocation of the enumerator and delegate in the next line